plugins/backend-ekg/src/Cardano/BM/Backend/Prometheus.lhs

```diff
@@ -42,28 +77,65 @@ spawnPrometheus ekg host port = Async.async $
-    renderSamples [] = mempty
-    renderSamples samples = mconcat
-        [ case sv of
-            Counter c -> renderNamedValue sk (int64Dec c)
-            Gauge g -> renderNamedValue sk (int64Dec g)
-            Label l -> if isFloat l
-                         then renderNamedValue sk (byteString $ encodeUtf8 l)
-                         else mempty
-            _ -> mempty
-        | (sk,sv) <- samples ]
-    renderNamedValue :: Text -> Builder -> Builder
-    renderNamedValue nm bld =
-        (byteString $ prepareName nm)
-        <> charUtf8 ' '
-        <> bld
-        <> charUtf8 '\n'
-    prepareName nm = encodeUtf8 $ replace " " "_" $ replace "-" "_" $ replace "." "_" nm
-    isFloat v = case double v of
-        Right (_n, "") -> True  -- only floating point number parsed, no leftover
-        _ -> False
 
+    -- rts.gc metrics are always here because they are predefined in ekg-core,
+    -- so we can group them.
+    extractRtsGcMetrics :: Text -> [(Text, Value)] -> MetricsGroup
+    extractRtsGcMetrics ns samples = MetricsGroup
+        { namespace = ns
+        , metrics =
+            [ case sv of
+                Counter c -> intMetric sk c
+                Gauge g   -> intMetric sk g
+                _         -> NoMetric -- rts.gc can contain Counter or Gauge only.
+            | (sk, sv) <- samples
+            ]
+        }
+      where
+        intMetric sk v =
+            Metric { mName  = maybe "" id $ T.stripPrefix (ns <> ".") sk
+                   , mType  = "int" -- All values are Int64.
+                   , mValue = NumberInt (fromIntegral v)
+                   }
+
+    -- We cannot make any assumptions about the format of 'sk' in other samples,
+    -- so group other samples into 'common' group.
+    extractOtherMetrics :: [(Text, Value)] -> MetricsGroup
+    extractOtherMetrics samples = MetricsGroup
+        { namespace = "common"
+        , metrics =
+            [ case sv of
+                Counter c -> mkMetric sk $ NumberInt (fromIntegral c)
+                Gauge g   -> mkMetric sk $ NumberInt (fromIntegral g)
+                Label l   -> case double l of
+                                 Left _       -> NoMetric
+                                 Right (r, _) -> mkMetric sk $ NumberReal r
+                _         -> NoMetric
+            | (sk, sv) <- samples
+            ]
+        }
+      where
+        mkMetric sk number =
+            let (withoutType, typeSuffix) = stripTypeSuffix sk number
+            in Metric { mName = withoutType, mType = typeSuffix, mValue = number }
+        stripTypeSuffix sk number =
+            if | ".us"   `T.isSuffixOf` sk -> doStrip sk ".us"
+               | ".ns"   `T.isSuffixOf` sk -> doStrip sk ".ns"
+               | ".s"    `T.isSuffixOf` sk -> doStrip sk ".s"
+               | ".B"    `T.isSuffixOf` sk -> doStrip sk ".B"
+               | ".int"  `T.isSuffixOf` sk -> doStrip sk ".int"
+               | ".real" `T.isSuffixOf` sk -> doStrip sk ".real"
```

Comment:
This looks like it's doing repeated traversal of the same text, which makes it quite costly.

I think this will be more efficient if we:

1. determine index of the last `.` character from the end
2. extract the tail, starting from that index, if any, and then compare it to the set, potentially dispatching on the first character (since it's different in all cases), to avoid comparing against the full set.

Another option would be to see if we just want to strip _all_ dotted suffixes -- that would be even more efficient.

Reply:
Thanks. Fixed.